### PR TITLE
More prep patches for miniso work

### PR DESCRIPTION
**Pull request by jlebon**
Split out of https://github.com/coreos/coreos-installer/pull/559. See individual commit messages.

### CODE REVIEW

**Comment by jlebon** [comment]
OK, clearly I need to updated my local clippy.

File: src/io.rs [hunk 1]
```diff
@@ -525,4 +570,31 @@ mod tests {
...
+        assert_eq!(data, outbuf);
+
+        // limit smaller than data
+        let mut outbuf: Vec<u8> = Vec::new();
+        let mut lim = LimitWriter::new(&mut outbuf, 90, "foo".into());
+        assert_eq!(
+            lim.write_all(&data).unwrap_err().to_string(),
+            "collision with foo at offset 90"
+        );
+    }
```
Comment (bgilbert) [inline src/io.rs:610]: There's some marginal value in testing multiple write calls per writer, particularly:
- checking that a write overrun succeeds up to the limit and then fails when `write` is called again at the limit,
- checking that a zero-length write succeeds at the limit.

**Review by bgilbert — APPROVED** [review]
Looks great; just some optional nits.  I like the `Address` solution.

**Comment by jlebon** [comment]
Updated!

**Review by bgilbert — APPROVED** [review]

**Pull request merged**